### PR TITLE
Fixed reconstruction when torch matrices are a secret and added tests to verify same 

### PR DESCRIPTION
**Pull request by kamathhrishi**
## Description
ReplicatedSharedTensor reconstruction for malicious security was failing when the secret was a torch tensor. This PR fixes the behavior and adds unit tests to verify the same. 

## How has this been tested?
- Added tests to verify the behavior

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

**Review by rasswanth-s — APPROVED** [review]
LGTM

File: src/sympc/tensor/replicatedshare_tensor.py [hunk 1]
```diff
@@ -304,10 +304,10 @@ def __reconstruct_malicious(
                 all_shares[(party_rank + 1) % (nparties)]
             )
 
-            if not value:
+            if (not isinstance(value, torch.Tensor)) and not value:
```
Comment (gmuraru) [inline src/sympc/tensor/replicatedshare_tensor.py:307]: MAYBE: Use `if value is None`

File: src/sympc/tensor/replicatedshare_tensor.py [hunk 2]
```diff
@@ -304,10 +304,10 @@ def __reconstruct_malicious(
...
                 value = share_sum
 
-            elif share_sum != value:
+            elif (share_sum != value).all():
```
Comment (gmuraru) [inline src/sympc/tensor/replicatedshare_tensor.py:310]: BUG: `any` not all

**Review by gmuraru — CHANGES_REQUESTED** [review]
See comments

**Review by rasswanth-s — APPROVED** [review]

**Comment by kamathhrishi** [comment]
> See comments

Made the required changes.

File: src/sympc/tensor/replicatedshare_tensor.py [hunk 3]
```diff
@@ -304,10 +304,9 @@ def __reconstruct_malicious(
                 all_shares[(party_rank + 1) % (nparties)]
             )
 
-            if not value:
+            if value is None:
                 value = share_sum
-
-            elif share_sum != value:
+            elif (share_sum != value).any():
```
Comment (aanurraj) [inline src/sympc/tensor/replicatedshare_tensor.py:309]: Q: Why you have .any() ?

File: tests/sympc/tensor/replicatedshare_tensor_test.py [hunk 4]
```diff
@@ -150,11 +150,15 @@ def test_rst_distribute_reconstruct(get_clients, parties, security) -> None:
     session = Session(protocol=protocol, parties=parties)
     SessionManager.setup_mpc(session)
 
-    secret = 42.32
+    # Test with torch matrix
+    secret1 = torch.ones([10, 10]) * 5
```
Comment (aanurraj) [inline tests/sympc/tensor/replicatedshare_tensor_test.py:154]: why not include zero and negative numbers  ? or a more mixed matrix?

File: tests/sympc/tensor/replicatedshare_tensor_test.py [hunk 5]
```diff
@@ -178,7 +182,7 @@ def test_invalid_malicious_reconstruction(get_clients, parties):
     session = Session(protocol=protocol, parties=parties)
     SessionManager.setup_mpc(session)
 
-    secret = 42.32
+    secret = torch.ones([10, 10])
```
Comment (aanurraj) [inline tests/sympc/tensor/replicatedshare_tensor_test.py:185]: same here too.

**Review by aanurraj — CHANGES_REQUESTED** [review]
check comments

File: src/sympc/tensor/replicatedshare_tensor.py (hunk 3 above)
Comment (kamathhrishi) [reply src/sympc/tensor/replicatedshare_tensor.py:309]: If any value in share_sum and value are not equal to each other it should evaluate to true.

File: tests/sympc/tensor/replicatedshare_tensor_test.py (hunk 4 above)
Comment (kamathhrishi) [reply tests/sympc/tensor/replicatedshare_tensor_test.py:154]: Sure.

**Comment by kamathhrishi** [comment]
> check comments

Made the required changes.

File: tests/sympc/tensor/replicatedshare_tensor_test.py [hunk 6]
```diff
@@ -150,11 +150,19 @@ def test_rst_distribute_reconstruct(get_clients, parties, security) -> None:
...
-    a = MPCTensor(secret=secret, session=session)
+    r1 = 0
+    r2 = 5
 
-    assert np.allclose(secret, a.reconstruct(), atol=1e-5)
+    secret1 = torch.FloatTensor(10, 10).uniform_(r1, r2)
+    a = MPCTensor(secret=secret1, session=session)
+    assert np.allclose(secret1, a.reconstruct(), atol=1e-3)
+
+    # Test with regular float values
```
Comment (gmuraru) [inline tests/sympc/tensor/replicatedshare_tensor_test.py:162]: Add another test for ```float``` vs ```matrix```

File: tests/sympc/tensor/replicatedshare_tensor_test.py [hunk 7]
```diff
@@ -178,8 +186,10 @@ def test_invalid_malicious_reconstruction(get_clients, parties):
     session = Session(protocol=protocol, parties=parties)
     SessionManager.setup_mpc(session)
 
-    secret = 42.32
+    r1 = 0
+    r2 = 5
 
+    secret = torch.FloatTensor(10, 10).uniform_(r1, r2)
```
Comment (gmuraru) [inline tests/sympc/tensor/replicatedshare_tensor_test.py:192]: Try to put some values here, such that we don't have random values in the test

**Comment by kamathhrishi** [comment]
@aanurraj and @gmuraru I have made the required changes.

**Review by gmuraru — APPROVED** [review]
LGTM!

**Review by aanurraj — APPROVED** [review]
